[quant] Add embeddingbag_prepack function that works on quantized tensor.

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42881 [quant] Add torchbind support for embedding_bag packed weights
* **#42762 [quant] Add embeddingbag_prepack function that works on quantized tensor.**

Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070632](https://our.internmc.facebook.com/intern/diff/D23070632)

supriyar committed:
[quant] Add embeddingbag_prepack function that works on quantized tensor.
Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -133,10 +134,73 @@ Tensor qembeddingbag_4bit_prepack(const Tensor& weight) {
+// TODO: Extend this to support 4-bits once 4-bit qtensor support is added.
+Tensor qembeddingbag_prepack(at::Tensor qweight) {
+  Tensor weight_contig = qweight.contiguous(qweight.suggest_memory_format());
+  const uint8_t* weight_data =
```
do we want to check the dtype here?

supriyar replied:
Sure, I can add a check. We don't check dtype in many ops though, probably because `data_ptr` would error out for incorrect type

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -133,10 +134,73 @@ Tensor qembeddingbag_4bit_prepack(const Tensor& weight) {
+
+  int64_t embedding_rows = qweight.size(0);
+  int64_t embedding_cols = qweight.size(1);
+  const auto qtype = qweight.qscheme();
```
nit: qscheme

jerryzh168 reviewed (approved):
Looks good

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070632](https://our.internmc.facebook.com/intern/diff/D23070632)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070632](https://our.internmc.facebook.com/intern/diff/D23070632)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070632](https://our.internmc.facebook.com/intern/diff/D23070632)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070632](https://our.internmc.facebook.com/intern/diff/D23070632)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add embeddingbag_prepack function that works on quantized tensor."


Summary:
Use a prepack function that accepts qtensor as an input. The output is a byte tensor with packed data.
This is currently implemented only for 8-bit. In the future once we add 4-bit support this function will be extended to support that too.

Note -In the following change I will add TorchBind support for this to support serialization of packed weights.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070632](https://our.internmc.facebook.com/intern/diff/D23070632)

[ghstack-poisoned]
